src/H5mpi.c

```diff
@@ -549,4 +549,238 @@ H5_mpio_create_large_type(hsize_t num_elements, MPI_Aint stride_bytes, MPI_Datat
+ *                            buffer. Analogous to MPI_(All)gatherv's
+ *                            `sendtype` parameter.
+ *
+ *              `recv_counts` - An array containing the number of elements
+ *                              to be received from each MPI rank.
+ *                              Analogous to MPI_(All)gatherv's `recvcount`
+ *                              parameter.
+ *
+ *              `displacements` - An array containing the displacements
+ *                                in the receive buffer where data from
+ *                                each MPI rank should be placed. Analogous
+ *                                to MPI_(All)gatherv's `displs` parameter.
+ *
+ *              `recv_type` - The MPI Datatype of the elements in the
+ *                            receive buffer. Analogous to
+ *                            MPI_(All)gatherv's `recvtype` parameter.
+ *
+ *              `allgather` - Specifies whether the gather operation to be
+ *                            performed should be MPI_Allgatherv (TRUE) or
+ *                            MPI_Gatherv (FALSE).
+ *
+ *              `root` - For MPI_Gatherv operations, specifies the rank
+ *                       that will receive the data sent by other ranks.
+ *                       Analogous to MPI_Gatherv's `root` parameter. For
+ *                       MPI_Allgatherv operations, this parameter is
+ *                       ignored.
+ *
+ *              `comm` - Specifies the MPI Communicator for the operation.
+ *                       Analogous to MPI_(All)gatherv's `comm` parameter.
+ *
+ *              `mpi_rank` - Specifies the calling rank's rank value, as
+ *                           obtained by calling MPI_Comm_rank on the
+ *                           MPI Communicator `comm`.
+ *
+ *              `mpi_size` - Specifies the MPI Communicator size, as
+ *                           obtained by calling MPI_Comm_size on the
+ *                           MPI Communicator `comm`.
+ *
+ *              `out_buf` - Resulting buffer that is allocated and
+ *                          returned to the caller after data has been
+ *                          gathered into it. Returned only to the rank
+ *                          specified by `root` for MPI_Gatherv
+ *                          operations, or to all ranks for
+ *                          MPI_Allgatherv operations.
+ *
+ *              `out_buf_num_entries` - The number of elements in the
+ *                                      resulting buffer, in terms of
+ *                                      the MPI Datatype provided for
+ *                                      `recv_type`.
+ *
+ * Notes:       This routine is collective across `comm`.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5_mpio_gatherv_alloc(void *send_buf, int send_count, MPI_Datatype send_type, const int recv_counts[],
```

Comment:
The two new functions here are simply wrappers around MPI_(All)gatherv that hide a bit of boilerplate code. Both allocate the receive buffer for the caller. The only difference between the two is that the "simple" function calculates the recv_counts and displacements arrays for the caller before making the MPI_(All)gatherv call.